source/extensions/filters/common/mutation_rules/mutation_rules.cc

```diff
@@ -0,0 +1,97 @@
+#include "source/extensions/filters/common/mutation_rules/mutation_rules.h"
+
+#include "envoy/http/header_map.h"
+
+#include "source/common/common/macros.h"
+#include "source/common/http/headers.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/match.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace MutationRules {
+
+using Http::LowerCaseString;
+
+class ExtraRoutingHeaders {
+public:
+  ExtraRoutingHeaders() {
+    const auto& hdrs = Http::Headers::get();
+    headers_.insert(hdrs.HostLegacy);
+    headers_.insert(hdrs.Host);
+    headers_.insert(hdrs.Method);
+    headers_.insert(hdrs.Scheme);
+  }
+
+  bool containsHeader(const LowerCaseString& name) const { return headers_.contains(name); }
+
+private:
+  absl::flat_hash_set<LowerCaseString> headers_;
+};
```

Comment:
This could just be a function returning an `absl::flat_hash_set`, but this works as well..